Xorg server removal plus grd

M4rtinK opened the pull request:
PR that replaces Tiger VNC with GNOME remote desktop running in RDP mode. 

At the moment built on top of @jexposit Xorg replacement PR & we will se how to actually land all this eventually.

pep8speaks commented:
Hello @M4rtinK! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:













* In the file [`pyanaconda/gnome_remote_destop.py`](https://github.com/rhinstaller/anaconda/blob/2ff67a4bdeed05770e220daed078c54e1d51ef3f/pyanaconda/gnome_remote_destop.py):

> [Line 33:1](https://github.com/rhinstaller/anaconda/blob/2ff67a4bdeed05770e220daed078c54e1d51ef3f/pyanaconda/gnome_remote_destop.py#L33): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 47:1](https://github.com/rhinstaller/anaconda/blob/2ff67a4bdeed05770e220daed078c54e1d51ef3f/pyanaconda/gnome_remote_destop.py#L47): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 81:100](https://github.com/rhinstaller/anaconda/blob/2ff67a4bdeed05770e220daed078c54e1d51ef3f/pyanaconda/gnome_remote_destop.py#L81): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (122 > 99 characters)
> [Line 113:100](https://github.com/rhinstaller/anaconda/blob/2ff67a4bdeed05770e220daed078c54e1d51ef3f/pyanaconda/gnome_remote_destop.py#L113): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (113 > 99 characters)
> [Line 164:100](https://github.com/rhinstaller/anaconda/blob/2ff67a4bdeed05770e220daed078c54e1d51ef3f/pyanaconda/gnome_remote_destop.py#L164): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (101 > 99 characters)

jexposit commented on pyanaconda/gnome_remote_destop.py:
```diff
@@ -0,0 +1,183 @@
+    def _start_grd_process(self):
+        """Start the GNOME remote desktop process."""
+        try:
+            execWithCapture(GRD_BINARY_PATH, [])
```
I think that you need to add the `--headless` param here:
https://gitlab.gnome.org/GNOME/gnome-remote-desktop/-/blob/master/data/gnome-remote-desktop-headless.service.in?ref_type=heads

Also, `execWithCapture` will block until the app starts. I believe you need to run it in the background.

M4rtinK replied:
Yep, running with `startProgram()` now, that does not wait for it to terminate. Still need to change the shutdown logic to account for that probably. :)

M4rtinK commented on pyanaconda/gnome_remote_destop.py:
```diff
@@ -0,0 +1,183 @@
+
+    As the server needs to be killed from the exit handler, it would have
+    to somehow get to the GRD instance. Like this, it can just kill
+    it by calling a function of the vnc module.
```
Fixed. :)

M4rtinK commented on pyanaconda/gnome_remote_destop.py:
```diff
@@ -0,0 +1,183 @@
+            util.ipmi_abort(scripts=self.anaconda.ksdata.scripts)
+            sys.exit(1)
+
+        # start by checking we have openssl available
```
Indeed, another copy paste error. :P Fixed. :)

M4rtinK commented on pyanaconda/display.py:
```diff
@@ -274,24 +276,19 @@ def setup_display(anaconda, options):
-    vnc_server.timeout = xtimeout
+    grd_server = GRDServer()  # The vnc Server object.
+    # pass the anaconda instance used for crash reporting
+    grd_server.anaconda = anaconda
```
Yeah, there might have been some obscure reason why the vnc module does it like this, but I don't think that applies there. So lets change it. :)

M4rtinK commented on pyanaconda/gnome_remote_destop.py:
```diff
@@ -0,0 +1,183 @@
+        # start by checking we have openssl available
+        if not os.path.exists(OPENSSL_BINARY_PATH):
+            stdoutLog.critical("No openssl binary found, can't generate certificates for GNOME remote desktop. Aborting.")
+            util.ipmi_abort(scripts=self.anaconda.ksdata.scripts)
```
Fixed by passing the anaconda instance to the constructor, as suggested. :)

M4rtinK commented on pyanaconda/gnome_remote_destop.py:
```diff
@@ -0,0 +1,183 @@
+        if not os.path.exists(GRD_BINARY_PATH):
+            # we assume there that the main binary being present implies grdctl is there as well
+            stdoutLog.critical("GNOME remote desktop tooling is not available. Aborting.")
+            util.ipmi_abort(scripts=self.anaconda.ksdata.scripts)
```
Fixed by passing the anaconda instance to the constructor, as suggested. :)

M4rtinK commented on pyanaconda/display.py:
```diff
@@ -200,6 +201,7 @@ def wl_preexec():
         signal.signal(signal.SIGINT, signal.SIG_IGN)
 
     argv = ["run-in-new-session.py",
+            "--headless",
```
Yep, should be already fixed, thanks! :)

M4rtinK replied:
Of course this also needs to be conditional eventually.

M4rtinK commented on pyanaconda/gnome_remote_destop.py:
```diff
@@ -0,0 +1,185 @@
+        # TODO: check openssl binary is available ?
+        execWithCapture(OPENSSL_BINARY_PATH,
+                        ["req", "-new",
+                         "-newkey" "rsa:4096",
```
Should have spotted that. :P Fixed. :)

M4rtinK commented on pyanaconda/gnome_remote_destop.py:
```diff
@@ -0,0 +1,185 @@
+        execWithCapture(OPENSSL_BINARY_PATH,
+                        ["req", "-new",
+                         "-newkey" "rsa:4096",
+                         "-days 720", "-nodes", "-x509",
```
Fixed. :)

M4rtinK commented on pyanaconda/gnome_remote_destop.py:
```diff
@@ -0,0 +1,185 @@
+        """Start the GNOME remote desktop process."""
+        try:
+            self.log.info("Starting GNOME remote desktop.")
+            self.proc = startProgram([GRD_BINARY_PATH, "--headless"])
```
That's a bit more tricky unfortunately, as due to the way this is done  now in the exit handler, it does not have an easy access to the object instance. But I'll figure something out. :)

M4rtinK replied:
So I've made the process object accessible to the exit handler & that seems to work fine to terminate the process at shutdown. :)

M4rtinK commented:
> I just realized that pipewire is not installed. I believe that it'll be a required dependency as well and maybe another service you need to initialize.

I've checked the image that gets generated & it contains `pipewire` & `pipewire-libs` as `gnome-remote-desktop` does have a dependency on `pipewire`.

But good point about the initialization, that could be the tricky part, as our installation environment is unfortunately quite different & cut down to the usual desktop environment in which pipewire is normally used. :P

Conan-Kudo commented:
> > I just realized that pipewire is not installed. I believe that it'll be a required dependency as well and maybe another service you need to initialize.
> 
> I've checked the image that gets generated & it contains `pipewire` & `pipewire-libs` as `gnome-remote-desktop` does have a dependency on `pipewire`.
> 
> But good point about the initialization, that could be the tricky part, as our installation environment is unfortunately quite different & cut down to the usual desktop environment in which pipewire is normally used. :P

You'll need my lorax change for this, since I had the same issue for Weston for similar reasons. https://github.com/weldr/lorax/pull/1364

jexposit commented:
Aha, that was why! Thanks Neal. Maybe we should reopen and merge your PR? (https://github.com/weldr/lorax/pull/1364)

Conan-Kudo commented:
For the moment, you should cherry-pick this from my PR: https://github.com/rhinstaller/anaconda/pull/5401/commits/ae3889c7db5d01950f514bd1caf5e32639faa82b

Conan-Kudo commented:
If the above doesn't work, let me know and I can revise the patch to lorax.

jexposit commented:
Hey @M4rtinK,

I tested your branch a bit more and you were missing a couple of things. [Here is the diff](https://github.com/rhinstaller/anaconda/files/14521998/g-r-d-working.diff.txt) to make g-r-d work :tada: 

I'm not sure if `Requires: gnome-keyring` is required for `--headless`, try to remove it, because I don't think we need it.

The main mystery I had to solve was related to the `$HOME` env var set to `/tmp`. I pruned it, but there might be a better solution, I'll let you investigate it.

M4rtinK commented:
> I tested your branch a bit more and you were missing a couple of things. [Here is the diff](https://github.com/rhinstaller/anaconda/files/14521998/g-r-d-working.diff.txt) to make g-r-d work 🎉

Nice, thanks! :)


> I'm not sure if `Requires: gnome-keyring` is required for `--headless`, try to remove it, because I don't think we need it.

Sure, I'll look into that. :)

> The main mystery I had to solve was related to the `$HOME` env var set to `/tmp`. I pruned it, but there might be a better solution, I'll let you investigate it.

Yep, I will see what can be done about that. We notoriously don't have a regular usable user account in the installation environment, though I'm more and more thinking about if this is not a problem, given how many things seem to expect one to exist these days. :P 

As we run everything as root anyway, I'll try what happens if `$HOME` is set to point to `/root`. :)

Conan-Kudo commented:
I'd turn it around and say maybe Anaconda should run as a dedicated user? We have polkit policies now, so we can also just make it auto-elevate properly, no?

M4rtinK commented:
/build-image --boot.iso

M4rtinK commented:
> I'd turn it around and say maybe Anaconda should run as a dedicated user? We have polkit policies now, so we can also just make it auto-elevate properly, no?

Yeah, even though we have so far managed to work around it somehow, the root-only, cut-down-on-file-level minimal image concept really seems to be getting more problematic by the day. :P 

It certainly made much more sense when it was first created in the ~ RHEL 7 timeframe, but seems really rather problematic in the modern day.

It would be good if we could move the image closer to the way modern distros are build, I am just afraid if we can still keep it minimal if we did that & if we actually have the capacity to pull it off. In any case, definitely not something we could tackle in this PR. :P

M4rtinK commented:
Substantial update for the PR:
* `inst.rdp`, `inst.rdp.username` and `inst.rdp.password` have been added & work
* interactive RDP username and password configuration in text mode is now possible
![TUI_RDP_question](https://github.com/rhinstaller/anaconda/assets/829558/c1526c5f-c6c8-4aa0-9b39-432a3d100e3a)
![TUI_RDP_question_1](https://github.com/rhinstaller/anaconda/assets/829558/49ecaac0-7875-419e-ace6-ce5e214dbdeb)
* VNC related boot options are still in place, but ineffective & just print a warning to the journal
![vnc_option_deprecation](https://github.com/rhinstaller/anaconda/assets/829558/40c0278a-fb45-4eb5-b4ad-9bdb446bf0d9)
![vnc_option_deprecation_1](https://github.com/rhinstaller/anaconda/assets/829558/527adf76-eaa8-4db2-bbaf-8e18e0a3a1d2)
* VNC options have been marked as deprecated in `anaconda --help`  output and in the boot option documentation
* the `inst.resolution` boot option can be used to set RDP virtual screen resolution

NOTE: VNC deprecation currently only targets RHEL, we have not yet decided how we will proceed with this once we port the Wayland migration patch se to Fedora, but will most likely want to keep RDP support in place, to keep RHEL feature parity.

There are also still some known issues/WiP:

* looks like gnome-kiosk tramples over `TTY1` with its stdout/stderr output
 * it is not easy to address due to the sophisticated `way run-in-new-session.py` starts the gnome-kiosk process & help of someone who knows more about those things might be needed
* there are apparently some more TUI overwriting/newline issues (see the screenshot demonstrating password entry), but this seems to happen also on a normal image when entering the VNC password

M4rtinK commented on anaconda.spec.in:
```diff
@@ -270,7 +270,9 @@ Requires: python3-meh-gui >= %{mehver}
 Requires: adwaita-icon-theme
 Requires: tecla
-Requires: tigervnc-server-minimal
+Requires: gnome-remote-desktop
```
Pipewire has been pulled in. :)

jexposit commented on pyanaconda/vnc.py:
```diff
@@ -1,303 +0,0 @@
-#
```
You might want to squash this commit ("Remove vnc.py") with "Replace VNC support with GNOME remote desktop". I think they make sense together.

M4rtinK replied:
Makes sense, lets do that. :)

jexposit commented on pyanaconda/display.py:
```diff
@@ -276,77 +319,55 @@ def setup_display(anaconda, options):
-            if len(cargs) > 1 and len(cargs[1]) > 0:
-                if len(cargs[1]) > 0:
-                    vnc_server.vncconnectport = cargs[1]
+        # FIXME: actually use user provided values
```
I think this FIXME is done

M4rtinK replied:
Indeed, removing it now. :)

jexposit commented on pyanaconda/display.py:
```diff
@@ -276,77 +319,55 @@ def setup_display(anaconda, options):
 
-    if options.vnc:
-        flags.usevnc = True
+    # FIXME: replace with RDP options
```
Done?

M4rtinK replied:
Yep.

jexposit commented on pyanaconda/display.py:
```diff
@@ -201,14 +224,34 @@ def wl_preexec():
+        if headless_resolution is None:
+            # use default value
+            headless_resolution = "1280x1024"
+        # FIXME: use inst.resolution for the virtual monitor
```
I believe `options.runres` is the right option and it should be working at the moment?

Conan-Kudo replied:
`1280x1024` is huge! Shouldn't this be `1024x768`?

jexposit commented on pyanaconda/display.py:
```diff
@@ -276,77 +319,55 @@ def setup_display(anaconda, options):
-                    "to use VNC mode instead?")
-        ask_vnc_question(anaconda, vnc_server, message)
+                    "to use remote graphical access via the RDP protocol instead?")
+        ask_vnc_question(anaconda, grd_server, message)
```
You might want to rename this method to `ask_rd_question` or similar.

M4rtinK replied:
Will do - there are actually a few remaining VNC references elsewhere I still like to remove/make generic. :)

jexposit reviewed (changes requested):
I added another batch of review comments. It is shaping up nicely. Let me know when the PR is in its final state and I'll review and test it again.

M4rtinK commented:
> I added another batch of review comments. It is shaping up nicely. Let me know when the PR is in its final state and I'll review and test it again.

Perfect, thanks, will do! :)

M4rtinK commented on anaconda.spec.in:
```diff
@@ -278,7 +278,11 @@ Requires: python3-meh-gui >= %{mehver}
 Requires: tecla
-Requires: tigervnc-server-minimal
+Requires: gnome-remote-desktop
+# FIXME: why does gnome-remote-desktop not drag gsm in when it needs it ?
```
Indeed, I think we can drop it, as we eventually tracked it down to Lorax cleaning up the so file, not a real dependency issue.

M4rtinK commented on pyanaconda/gnome_remote_destop.py:
```diff
@@ -0,0 +1,223 @@
+        # then create folder for the certs
+        os.makedirs(GRD_RDP_CERT_DIR)
+        # generate the certs
+        # TODO: check openssl binary is available ?
```
Yep, lets remove this line. :)

Conan-Kudo commented:
I would prefer we didn't completely gut the VNC support. I'd like to revive my Weston patch for consideration to be supported in Anaconda boot.iso and being able to plumb all these things through with the existing infrastructure would be nice.

(I can also use the new plumbing for RDP too, since Weston supports both.)

M4rtinK commented:
> I would prefer we didn't completely gut the VNC support. I'd like to revive my Weston patch for consideration to be supported in Anaconda boot.iso and being able to plumb all these things through with the existing infrastructure would be nice.
> 
> (I can also use the new plumbing for RDP too, since Weston supports both.)

I'm actually making the code a bit generic in places (eq. all those `rd` & remote desktop references) in case we might decide to keep VNC support in place in Fedora. I think it might be less hassle to just introduce the RDP options rather than drop VNC and introduce RDP support at the same time. 

But this is not set in stone right now, just an idea.

Conan-Kudo commented:
> > I would prefer we didn't completely gut the VNC support. I'd like to revive my Weston patch for consideration to be supported in Anaconda boot.iso and being able to plumb all these things through with the existing infrastructure would be nice.
> > (I can also use the new plumbing for RDP too, since Weston supports both.)
> 
> I'm actually making the code a bit generic in places (eq. all those `rd` & remote desktop references) in case we might decide to keep VNC support in place in Fedora. I think it might be less hassle to just introduce the RDP options rather than drop VNC and introduce RDP support at the same time.
> 
> But this is not set in stone right now, just an idea.

It's a very good one! Let's do it!

jadahl commented on pyanaconda/display.py:
```diff
@@ -210,41 +207,67 @@ def do_startup_x11_actions():
+        if headless_resolution is None:
+            # use default value
+            headless_resolution = "1280x1024"
+        argv.extend(["--headless", "--virtual-monitor", headless_resolution])
```
Running the headless g-r-d should automatically create its own virtual monitor, with a resolution that matches the RDP clients.

M4rtinK replied:
I seem to remember it would not work without us setting the resolution with this option, but I'll re-check.

M4rtinK replied:
BTW, if I drop the `--virtual-monitor` option in the headless mode, it does not seem to be happy at all:
![no_virtual_monitor](https://github.com/rhinstaller/anaconda/assets/829558/44115c03-cd04-4b9e-a85b-4f223cdeb801)

All this before it becomes possible to connect - I guess it default to 0x0 pixels resolution or something ?

The Anaconda GUI needs to be running (and might even do things by itself as IIRC it should be possible to connect to an automated kickstart installation running in GUI mode) before the first client connects, so some sensible default "headless" screen size should be used, to be possibly overridden by the clients connecting.

IIRC @jstodola mentioned that the GUI already does handle screen resolution changes in some VM installation use cases.

Still, I wonder what would happen if a second user connects (which seems to be possible with RDP) & requests a different screen resolution ?

jadahl replied:
With GNOME 45 / F39, I think you need to tweak the gsettings (`gsettings set org.gnome.desktop.remote-desktop.rdp screen-share-mode extend`), but in GNOME 46 / F40 it'll be automatic.

jadahl replied:
There might be bugs in clients that doesn't handle the case where they try to show up before the RDP client connects, though (the warnings seem to be about that).

jadahl commented on pyanaconda/display.py:
```diff
@@ -210,41 +207,67 @@ def do_startup_x11_actions():
+    raise TimeoutError("Timeout trying to start gnome-kiosk")
+
+
+def set_resolution(runres):
```
This doesn't look necessary, since it should be handled via the RDP protocol.

jexposit replied:
The method is only used to set the screen resolution in the non-headless session. It should be safe to keep calling it:

```python
            if options.runres and anaconda.gui_mode and not flags.use_rd:
                def on_mutter_ready(observer):
                    set_resolution(options.runres)
                    observer.disconnect()
```

jadahl replied:
ah, I missed the `not` before `flags.use_rd`.

jadahl replied:
I wonder if it'd not be better to generate an monitors.xml file instead, as that'd avoid mutter first setting one resolution, then changing to the one here. It would require copying the way mutter decides how to index port numbers (e.g. DP-#).

Conan-Kudo replied:
That's just libdisplay-info, no? IIRC, most Wayland compositors use that to enumerate outputs.

jadahl replied:
No, the enumeration comes from the KMS uAPI. libdisplay-info is to get info about the display e.g. colorimetry information and stuff like that. In mutter the connector name comes from here: https://gitlab.gnome.org/GNOME/mutter/-/blob/e4dd50e1dd7e6a741eb645ec5526768e33082259/src/backends/native/meta-kms-connector.c#L1562-1593 which effectively is a prefix (e.g. `DP-` or `HDMI-` and the an ID from KMS). That isn't really API though, maybe it'll one day try to be smarter.

M4rtinK replied:
BTW, if I drop the `--virtual-monitor` option in the headless mode, it does not seem to be happy at all:
![no_virtual_monitor](https://github.com/rhinstaller/anaconda/assets/829558/44115c03-cd04-4b9e-a85b-4f223cdeb801)

All this before it becomes possible to connect - I guess it default to 0x0 pixels resolution or something ?

jadahl commented on pyanaconda/gnome_remote_destop.py:
```diff
@@ -0,0 +1,226 @@
+        try:
+            self.log.info("Starting GNOME remote desktop.")
+            global grd_process
+            grd_process = startProgram([GRD_BINARY_PATH, "--headless"],
```
Any reason to not use the systemd service (`gnome-remote-desktop-headless.service`)?

M4rtinK replied:
Is it available in some released gnome-remote-desktop package ?

I have seen it here: https://gitlab.gnome.org/GNOME/gnome-remote-desktop/-/merge_requests/215/diffs

And that looks like it has been merged, but I only see `/usr/lib/systemd/user/gnome-remote-desktop.service` in `gnome-remote-desktop-45.1-1` that I have installed on my Fedora 39 machine and latest Rawhide seems to have justa a rebuild of that & there is indeed no such file in the RPM: https://kojipkgs.fedoraproject.org/compose/40/Fedora-40-20240319.1/compose/Everything/x86_64/os/Packages/g/gnome-remote-desktop-45.1-4.fc40.x86_64.rpm

So maybe the MR introducing this is merged but not yet packaged ?

jadahl replied:
It seems so yes, merged but not packaged. Seems to be in progress (https://src.fedoraproject.org/rpms/gnome-remote-desktop/pull-request/6) - I'll see what is left to do.

M4rtinK replied:
OK, thanks! :) We can switch to that service later when it becomes available in the environment.

M4rtinK replied:
Maybe the spam with `--headless` and the `gnome-remote-desktop-headless.service` are related ? Because I guess that if the GNOME remote desktop binary was launched by system then maybe it would capture **all** of its output & it will end up in Journal (which we would like) instead of on TTY1 ?

Thanks to @jstodola or possibly @rvykydal for the idea, not sure which right now. :)

M4rtinK replied:
I've created a bug in Bugzilla, so we can track the missing service file. :)
https://bugzilla.redhat.com/show_bug.cgi?id=2270718

M4rtinK commented:
/kickstart-test --testtype smoke

M4rtinK commented:
/kickstart-test --testtype ui

M4rtinK commented:
/kickstart-test --skip-testtypes knownfailure,manual,skip-on-fedora,gh576,gh595,gh640,gh641,gh680,gh740,gh769,gh774,gh777,gh910,gh890,gh871,rhbz1853668,gh975,gh1023

M4rtinK commented:
/kickstart-test user-multiple

rvykydal commented:
/kickstart-test user-multiple

M4rtinK commented:
/kickstart-test user-multiple

M4rtinK commented:
/kickstart-test --testtype users

M4rtinK commented:
Looks like what we are seeing with the users-multiple test failing is a race condition - many tests use the `keyboard` command that likely triggers this, yet were unaffected in that run. So lets try again. ;-)

M4rtinK commented:
/kickstart-test --skip-testtypes knownfailure,manual,skip-on-fedora,gh576,gh595,gh640,gh641,gh680,gh740,gh769,gh774,gh777,gh910,gh890,gh871,rhbz1853668,gh975,gh1023

M4rtinK commented:
/kickstart-test --skip-testtypes knownfailure,manual,skip-on-fedora,gh576,gh595,gh640,gh641,gh680,gh740,gh769,gh774,gh777,gh910,gh890,gh871,rhbz1853668,gh975,gh1023

M4rtinK commented:
/kickstart-test --skip-testtypes knownfailure,manual,skip-on-fedora,gh576,gh595,gh640,gh641,gh680,gh740,gh769,gh774,gh777,gh910,gh890,gh871,rhbz1853668,gh975,gh1023

M4rtinK commented:
/kickstart-test default-systemd-target-vnc,default-fstype,ui_vnc,rpm-ostree-container-silverblue,ignoredisk-2

M4rtinK commented:
/kickstart-test default-systemd-target-vnc default-fstype ui_vnc,rpm-ostree-container-silverblue ignoredisk-2

M4rtinK commented:
Rebased on the latest version of the Xorg removal PR (#5485), which now also means rebasing on RHEL 10 branch. :)

M4rtinK commented:
/kickstart-test --testtype smoke

M4rtinK commented:
/kickstart-test --testtype smoke

M4rtinK commented:
/build-image --boot.iso

M4rtinK commented:
/build-image --boot.iso

M4rtinK commented:
/kickstart-test --testtype smoke

M4rtinK commented:
/kickstart-test --testtype smoke

M4rtinK commented:
/build-image --boot.iso

rvykydal commented on pyanaconda/modules/localization/gk_keyboard_manager.py:
```diff
@@ -0,0 +1,135 @@
+        """Transforms a layout path as returned by GNOME Kiosk to "layout (variant)".
+
+        :param layout_path: D-Bus path to the layout.
+            (e.g. "/org/gnome/Kiosk/InputSources/xkb_cz_2b_mon_5f_todo_5f_galik")
```
a typo "_cz_" -> "_cn_"

M4rtinK commented:
/build-image --boot.iso

M4rtinK commented:
/build-image --boot.iso

M4rtinK commented:
/kickstart-test --skip-testtypes knownfailure,manual,skip-on-rhel,skip-on-rhel-10,gh576,gh640,gh804,gh1090,gh1104,gh1106,gh1105,gh1108,gh1109,gh1107,gh1110

M4rtinK commented:
/kickstart-test --skip-testtypes knownfailure,manual,skip-on-rhel,skip-on-rhel-10,gh576,gh640,gh804,gh1090,gh1104,gh1106,gh1105,gh1108,gh1109,gh1107,gh1110

jexposit commented:
Hey @M4rtinK, check @jkonecny12's comment, please: https://github.com/rhinstaller/anaconda/pull/5485#issuecomment-2126894245

The reference to https://issues.redhat.com/browse/RHEL-38407 needs to be added here. Rebase on top of the `xorg-server-removal`, I just added the reference to the other ticket.

jkonecny12 commented on docs/boot-options.rst:
```diff
@@ -525,13 +555,20 @@ may connect.
 Set a password on the VNC server used by the installer.
 
+This option is deprecated and will be removed in future releases.
+Use ``inst.rdp`` and related boot options instead.
```
I think here should be ``inst.rdp.password`` instead.

M4rtinK replied:
Indeed, fixed. :)

jkonecny12 commented on pyanaconda/core/constants.py:
```diff
@@ -84,7 +84,7 @@
 
 # VNC questions
-USEVNC = N_("Start VNC")
+USERDP = N_("Use remote graphical mode")
```
I think we should mention here RDP with capital letters ``Use Remote Desktop Protocol graphical mode``. My point is that it should clarify what user should use to do the connection to the installation.

jkonecny12 commented on pyanaconda/ui/tui/spokes/askrd.py:
```diff
@@ -114,48 +135,50 @@ def apply(self):
     @property
     def completed(self):
-        return True # We're always complete
+        return True  # We're always complete
```
Don't you want to check if username is correctly set together with password? What will happen if user will miss `username`?

M4rtinK replied:
So I've just tried it - it tells you that you have not specified and username & asks again. :)

![no_username](https://github.com/rhinstaller/anaconda/assets/829558/109dcc38-720a-4ebd-84a9-1f409b9fbe40)

So this should be covered. :)

M4rtinK commented on anaconda.spec.in:
```diff
@@ -252,6 +252,13 @@ Requires: gsettings-desktop-schemas
+Requires: gnome-remote-desktop
+# needed to generate RDP certs at runtime
+Requires: openssl
+# FIXME: needed by GNOME kisok but not declared ?
```
Fixed, thanks! :)

M4rtinK commented:
Rebased PR on latest version of https://github.com/rhinstaller/anaconda/pull/5485 & added correct Jira references. Also dropped python-pam from the temporary COPR repo, as it should be finally in the RHEL 10 repos.

M4rtinK commented:
/kickstart-test --skip-testtypes knownfailure,manual,skip-on-rhel,skip-on-rhel-10,gh576,gh640,gh804,gh1090,gh1104,gh1106,gh1105,gh1108,gh1109,gh1107,gh1110

jstodola commented:
This change should also be documented in `docs/release-notes.rst`.

M4rtinK commented:
/kickstart-test --testtype smoke

jkonecny12 commented on docs/boot-options.rst:
```diff
@@ -514,6 +514,36 @@ Specify screen size for the installer. Use format nxm, where n is the
+
+Multiple RDP clients can connect.
+
+When using ``inst.rdp``, you also need to set RDP username and password using the
+``inst.rdp.username`` and ``inst.rdp.password`` boot options.
+
+.. inst.rdp.username:
+
+inst.rdp.username
+^^^^^^^^^^^^^^^^^
+
+Set username for the RDP session. To enable RDP access, also use the
+``inst.rdp`` and ``inst.rdp.password`` boot options.
+
+.. inst.rdp.password:
+
+inst.rdp.password
+^^^^^^^^^^^^^^^^^
+
+Set password for the RDP session. To enable RDP access, also use the
+``inst.rdp`` and ``inst.rdp.username`` boot options.
```
I wonder if we need user to specify all of these options. I mean what would be the use-case for having `inst.rdp` without password and username -> invalid. On the same note what would be the use case for user to have `inst.rdp.username` without `inst.rdp`.

M4rtinK replied:
So this is what happens if you use just `inst.rdp`:
![inst rdp_only](https://github.com/rhinstaller/anaconda/assets/829558/f34e7d41-7bed-4005-8e93-393000e6b61f)
If you use just `inst.username`, it boots to GUI without warning - I think that's wrong & I'll fix it.
If you use `inst.rdp` and `inst.username`, you get:
![inst rdp_inst username](https://github.com/rhinstaller/anaconda/assets/829558/101a3019-2963-4280-89ec-d8443e5374ac)
If you use just `inst.password`, it boots to GUI without warning - I think that's wrong & should be fixed as well.
If you use `inst.rdp` and `inst.password`, you get:
![inst rdp_inst password](https://github.com/rhinstaller/anaconda/assets/829558/36e8fc75-c85c-497c-8289-339a2412d9bf)

So indeed, this should be fixed up and simplified. I suggest:

- we drop the lone `inst.rdp` from docs & make sure there is an error when it is used (as we can't avoid people just typing that into the boot options)
- show an error when only `inst.password` or `inst.username` is used

@jkonecny12 @jstodola does it make sense like this ? :)

jstodola replied:
The use case for just `inst.rdp` is to be able to specify the username and password interactively. You might not want to specify the password on the kernel command line. IMO, it makes sense to keep the boot option.

Looking at the current `inst.vnc` and `inst.vncpassword` behavior,  having just `inst.vncpassword` itself on the kernel command line doesn't start VNC mode. So, my suggestion is to keep this logic and require `inst.rdp` for remote graphical installations.

Perhaps if `inst.rdp` is specified on the kernel cmdline and one of `inst.rdp.username` or `inst.rdp.password` is missing, anaconda should ask for the missing value interactively, while respecting the other value specified on the kernel cmdline.

M4rtinK replied:
I like that idea! I've adjusted the interactive code accordingly:

- if `inst.rdp` is used, go directly to the interactive entry of RDP credentials
- if username or password is set via boot option, ask just for the other bit
- only using `inst.rdp.username` or `inst.rdp.password` without also `inst.rdp` should not do anything, same as with VNC so far
- also some more minor cleanups and UX improvements

![better_interactive_auth](https://github.com/rhinstaller/anaconda/assets/829558/0cbbd59f-51ea-4a8a-9950-2a431b91e556)

You might notice some other stuff writing over the console - definitely something to investigate but IMHO not blocking.

I've created also a testing image - just boot the first boot options & you should go straight to the interactive entry:
https://fedorapeople.org/groups/anaconda/wayland/gnome_remote_desktop/better_rdp_interactive_auth_mk1.iso

jkonecny12 commented on pyanaconda/display.py:
```diff
@@ -106,72 +105,81 @@ def start_spice_vd_agent():
     loop = App.get_event_loop()
     loop.set_quit_callback(tui_quit_callback)
-    spoke = AskVNCSpoke(anaconda.ksdata, message=message)
+    spoke = AskRDSpoke(anaconda.ksdata, message=message)
```
This class do not exists in this commit. The commit is broken and fixed by the next one.
Please fix this issue.

M4rtinK replied:
Changed the commit order. :)

jkonecny12 reviewed (changes requested):
Please fix the broken commit.

Also I don't see almost no test coverage. It would be great to have at least something for the new functionality.

M4rtinK commented:
> Also I don't see almost no test coverage. It would be great to have at least something for the new functionality.

There is a kickstart tests PR addressing this ready to be merged once the functionality is in: https://github.com/rhinstaller/kickstart-tests/pull/1175

M4rtinK commented on anaconda.spec.in:
```diff
@@ -256,6 +256,13 @@ Requires: gsettings-desktop-schemas
 Requires: nm-connection-editor
 Requires: librsvg2
 Requires: gnome-kiosk
+# FIXME: move GNOME remote desktop deps to Lorax templates to avoid
+# pulling them in every time anaconda-gui is installed
+Requires: gnome-remote-desktop
```
Yeah, I think its a leftover and we can drop it. :)

M4rtinK replied:
Fixed & I've also replaced the other FIXME for about `mesa-dri-dirvers` with a note about how the missing dependency is intentional according to the maintainers. :)

M4rtinK replied:
/kickstart-test --skip-testtypes knownfailure,manual,skip-on-rhel,skip-on-rhel-10,gh576,gh640,gh804,gh1090,gh1104,gh1106,gh1105,gh1108,gh1109,gh1107,gh1110

jkonecny12 commented on pyanaconda/ui/tui/spokes/askrd.py:
```diff
@@ -114,49 +135,83 @@ def apply(self):
+        if password is not None:
+            self._password = password
+        else:
+           self._password = ""
```
Please fix the indentation.

jkonecny12 reviewed (changes requested):
Please fix the indentation. Otherwise it looks good.

M4rtinK commented:
/kickstart-test --skip-testtypes knownfailure,manual,skip-on-rhel,skip-on-rhel-10,gh576,gh640,gh804,gh1090,gh1104,gh1106,gh1105,gh1108,gh1109,gh1107,gh1110

jkonecny12 reviewed (approved):
Looks good to me now.

M4rtinK commented:
/kickstart-test --testtype smoke
